### ISSUE TITLE
AWS ECS container metrics receiver panics on ECS Anywhere

### ISSUE THREAD

**Issue by bpalermo** [aws-observability/aws-otel-collector#2110]
**Describe the bug**
AWS ECS container metrics receiver panics on ECS Anywhere

**Steps to reproduce**
Enable receiver/awsecscontainermetricsreceiver

**What did you expect to see?**
No panic

**What did you see instead?**
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2ac4421]

goroutine 97 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics.getContainerMetrics(0xc000944f80, 0xc0002b3200?)
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.76.3/internal/awsecscontainermetrics/metrics_helper.go:27 +0xa1
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics.convertContainerMetrics(_, _, {{0xc000711a00, 0x7e}, {0xc000abb0c0, 0xb}, {0xc000745a20, 0x1e}, {0xc000731c80, 0x40}, ...})
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.76.3/internal/awsecscontainermetrics/accumulator.go:80 +0x58
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics.(*metricDataAccumulator).getMetricsData(0xc00092dcf8, 0x9?, {{0x0, 0x0}, {0xc000abb090, 0xb}, {0xc000689400, 0x3, 0x4}, {0xc0006f45b8, ...}, ...}, ...)
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.76.3/internal/awsecscontainermetrics/accumulator.go:51 +0x2e5
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics.MetricsData(...)
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.76.3/internal/awsecscontainermetrics/metrics.go:27
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver.(*awsEcsContainerMetricsReceiver).collectDataFromEndpoint(0xc000cf1f40, {0x4b5e238, 0xc000317d10})
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.76.3/receiver.go:99 +0x3e9
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver.(*awsEcsContainerMetricsReceiver).Start.func1()
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.76.3/receiver.go:71 +0xa9
created by github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver.(*awsEcsContainerMetricsReceiver).Start
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.76.3/receiver.go:64 +0xca
```
**Environment**
aws-otel-collector v0.29.1
ecs-agent 1.71.1

**Additional context**

**Comment by ziggythehamster** [comment]
I am experiencing this on regular ECS as well, on aws-otel-collector v0.30.0:

```
2023/06/22 05:50:30 ADOT Collector version: v0.30.0
--
2023/06/22 05:50:30 found no extra config, skip it, err: open /opt/aws/aws-otel-collector/etc/extracfg.txt: no such file or directory
2023/06/22 05:50:30 attn: users of the prometheus receiver, prometheus exporter or prometheusremotewrite exporter please refer to https://github.com/aws-observability/aws-otel-collector/issues/2043 in regards to an ADOT Collector v0.31.0 breaking change
2023/06/22 05:50:30 Reading AOT config from environment: "exporters":
"prometheusremotewrite":
"auth":
"authenticator": "sigv4auth"
"endpoint": "redacted"
"extensions":
"health_check":
"check_collector_pipeline":
"enabled": true
"endpoint": "0.0.0.0:13133"
"path": "/"
"sigv4auth":
"assume_role":
"arn": "redacted"
"sts_region": "us-west-2"
"processors":
"attributes":
"actions":
- "action": "upsert"
"key": "cluster"
"value": "redacted"
- "action": "upsert"
"key": "__replica__"
"value": "$HOSTNAME"
"receivers":
"awsecscontainermetrics":
"collection_interval": "20s"
"service":
"extensions":
- "health_check"
- "sigv4auth"
"pipelines":
"metrics":
"exporters":
- "prometheusremotewrite"
"processors":
- "attributes"
"receivers":
- "awsecscontainermetrics"
2023-06-22T05:50:30.219Z	info	service/telemetry.go:104	Setting up own telemetry...
2023-06-22T05:50:30.221Z	info	service/telemetry.go:127	Serving Prometheus metrics	{     "address": ":8888",     "level": "Basic" }
2023-06-22T05:50:30.231Z	info	service/service.go:131	Starting aws-otel-collector...	{     "Version": "v0.30.0",     "NumCPU": 2 }
2023-06-22T05:50:30.231Z	info	extensions/extensions.go:30	Starting extensions...
2023-06-22T05:50:30.231Z	info	extensions/extensions.go:33	Extension is starting...	{     "kind": "extension",     "name": "sigv4auth" }
2023-06-22T05:50:30.231Z	info	extensions/extensions.go:37	Extension started.	{     "kind": "extension",     "name": "sigv4auth" }
2023-06-22T05:50:30.231Z	info	extensions/extensions.go:33	Extension is starting...	{     "kind": "extension",     "name": "health_check" }
2023-06-22T05:50:30.231Z	info	healthcheckextension@v0.78.0/healthcheckextension.go:34	Starting health_check extension	{     "kind": "extension",     "name": "health_check",     "config": {         "Endpoint": "0.0.0.0:13133",         "TLSSetting": null,         "CORS": null,         "Auth": null,         "MaxRequestBodySize": 0,         "IncludeMetadata": false,         "Path": "/",         "ResponseBody": null,         "CheckCollectorPipeline": {             "Enabled": true,             "Interval": "5m",             "ExporterFailureThreshold": 5         }     } }
2023-06-22T05:50:30.232Z	warn	internal/warning.go:40	Using the 0.0.0.0 address exposes this server to every network interface, which may facilitate Denial of Service attacks	{     "kind": "extension",     "name": "health_check",     "documentation": "https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/security-best-practices.md#safeguards-against-denial-of-service-attacks" }
2023-06-22T05:50:30.253Z	info	extensions/extensions.go:37	Extension started.	{     "kind": "extension",     "name": "health_check" }
2023-06-22T05:50:30.253Z	info	healthcheck/handler.go:129	Health Check state change	{     "kind": "extension",     "name": "health_check",     "status": "ready" }
2023-06-22T05:50:30.253Z	info	service/service.go:148	Everything is ready. Begin running and processing data.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2b23941]
goroutine 101 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics.getContainerMetrics(0xc0000dc200, 0xc000324e00?)
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.78.0/internal/awsecscontainermetrics/metrics_helper.go:16 +0xa1
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics.convertContainerMetrics(_, _, {{0xc000ca2300, 0x7d}, {0xc000d26bd0, 0x17}, {0xc000c9a680, 0x1e}, {0xc0000dfbc0, 0x40}, ...})
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.78.0/internal/awsecscontainermetrics/accumulator.go:69 +0x58
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics.(*metricDataAccumulator).getMetricsData(0xc000b17cf8, 0x0?, {{0xc000cbf640, 0xa}, {0xc000cbf600, 0xa}, {0xc000827400, 0x1, 0x4}, {0xc000cb4870, ...}, ...}, ...)
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.78.0/internal/awsecscontainermetrics/accumulator.go:40 +0x2e5
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics.MetricsData(...)
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.78.0/internal/awsecscontainermetrics/metrics.go:16
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver.(*awsEcsContainerMetricsReceiver).collectDataFromEndpoint(0xc0004ce840, {0x4c352b8, 0xc000cb6230})
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.78.0/receiver.go:88 +0x3e9
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver.(*awsEcsContainerMetricsReceiver).Start.func1()
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.78.0/receiver.go:60 +0xa9
created by github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver.(*awsEcsContainerMetricsReceiver).Start
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver@v0.78.0/receiver.go:53 +0xca
```

**Comment by ziggythehamster** [comment]
Looking at the code, I think I found the bug.

In `opentelemetry-collector-contrib/receiver/awsecscontainermetricsreceiver/internal/awsecscontainermetrics/metrics_helper.go`, there's a check for `stats.Memory.Stats` not being nil, but there is no check that `stats.Memory.Stats["cache"]` is present and not nil:

```go
		if stats.Memory.Stats != nil {
			m.MemoryUtilized = (*stats.Memory.Usage - stats.Memory.Stats["cache"]) / bytesInMiB
		}
```

...but I don't have a cache entry in my stats:

```json
{
  "read": "2023-06-22T06:57:11.271562027Z",
  "preread": "2023-06-22T06:57:10.267702451Z",
  "pids_stats": {
    "current": 8,
    "limit": 48137
  },
  "blkio_stats": {
    "io_service_bytes_recursive": null,
    "io_serviced_recursive": null,
    "io_queue_recursive": null,
    "io_service_time_recursive": null,
    "io_wait_time_recursive": null,
    "io_merged_recursive": null,
    "io_time_recursive": null,
    "sectors_recursive": null
  },
  "num_procs": 0,
  "storage_stats": {},
  "cpu_stats": {
    "cpu_usage": {
      "total_usage": 1229675000,
      "usage_in_kernelmode": 1114129000,
      "usage_in_usermode": 115545000
    },
    "system_cpu_usage": 13041500000000,
    "online_cpus": 2,
    "throttling_data": {
      "periods": 0,
      "throttled_periods": 0,
      "throttled_time": 0
    }
  },
  "precpu_stats": {
    "cpu_usage": {
      "total_usage": 1229526000,
      "usage_in_kernelmode": 1113993000,
      "usage_in_usermode": 115533000
    },
    "system_cpu_usage": 13039500000000,
    "online_cpus": 2,
    "throttling_data": {
      "periods": 0,
      "throttled_periods": 0,
      "throttled_time": 0
    }
  },
  "memory_stats": {
    "usage": 46141440,
    "stats": {
      "active_anon": 8192,
      "active_file": 212992,
      "anon": 18165760,
      "anon_thp": 10485760,
      "file": 26955776,
      "file_dirty": 0,
      "file_mapped": 26284032,
      "file_writeback": 0,
      "inactive_anon": 18161664,
      "inactive_file": 26738688,
      "kernel_stack": 131072,
      "pgactivate": 52,
      "pgdeactivate": 0,
      "pgfault": 17142,
      "pglazyfree": 0,
      "pglazyfreed": 0,
      "pgmajfault": 6581,
      "pgrefill": 0,
      "pgscan": 0,
      "pgsteal": 0,
      "shmem": 4096,
      "slab": 468784,
      "slab_reclaimable": 296288,
      "slab_unreclaimable": 172496,
      "sock": 0,
      "thp_collapse_alloc": 0,
      "thp_fault_alloc": 10,
      "unevictable": 0,
      "workingset_activate": 0,
      "workingset_nodereclaim": 0,
      "workingset_refault": 0
    },
    "limit": 1073741824
  },
  "name": "/redacted",
  "id": "redacted",
  "networks": {
    "eth0": {
      "rx_bytes": 12392,
      "rx_packets": 42,
      "rx_errors": 0,
      "rx_dropped": 0,
      "tx_bytes": 3947,
      "tx_packets": 21,
      "tx_errors": 0,
      "tx_dropped": 0
    }
  },
  "network_rate_stats": {
    "rx_bytes_per_sec": 0,
    "tx_bytes_per_sec": 0
  }
}
```

Other useful info:

```
# uname -a
Linux redacted 4.18.0-477.13.1.el8_8.x86_64 #1 SMP Tue May 30 22:15:39 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

# cat /proc/vmstat
nr_free_pages 1456879
nr_zone_inactive_anon 107899
nr_zone_active_anon 8974
nr_zone_inactive_file 13489
nr_zone_active_file 151311
nr_zone_unevictable 0
nr_zone_write_pending 362
nr_mlock 0
nr_bounce 0
nr_zspages 0
nr_free_cma 0
numa_hit 42273680
numa_miss 0
numa_foreign 0
numa_interleave 46025
numa_local 42273680
numa_other 0
nr_inactive_anon 107899
nr_active_anon 8974
nr_inactive_file 13489
nr_active_file 151311
nr_unevictable 0
nr_slab_reclaimable 20573
nr_slab_unreclaimable 45474
nr_isolated_anon 0
nr_isolated_file 0
workingset_nodes 0
workingset_refault_anon 0
workingset_refault_file 0
workingset_activate_anon 0
workingset_activate_file 0
workingset_restore_anon 0
workingset_restore_file 0
workingset_nodereclaim 0
nr_anon_pages 90543
nr_mapped 70374
nr_file_pages 183447
nr_dirty 362
nr_writeback 0
nr_writeback_temp 0
nr_shmem 18647
nr_shmem_hugepages 0
nr_shmem_pmdmapped 0
nr_file_hugepages 0
nr_file_pmdmapped 0
nr_anon_transparent_hugepages 81
nr_vmscan_write 0
nr_vmscan_immediate_reclaim 0
nr_dirtied 67267
nr_written 61585
nr_throttled_written 0
nr_kernel_misc_reclaimable 0
nr_foll_pin_acquired 0
nr_foll_pin_released 0
nr_kernel_stack 4752
nr_page_table_pages 2070
nr_swapcached 0
nr_dirty_threshold 305051
nr_dirty_background_threshold 152339
pgpgin 1598423
pgpgout 4316508
pswpin 0
pswpout 0
pgalloc_dma 256
pgalloc_dma32 256
pgalloc_normal 56079595
pgalloc_movable 0
allocstall_dma 0
allocstall_dma32 0
allocstall_normal 0
allocstall_movable 0
pgskip_dma 0
pgskip_dma32 0
pgskip_normal 0
pgskip_movable 0
pgfree 57537346
pgactivate 375318
pgdeactivate 0
pglazyfree 0
pgfault 50833599
pgmajfault 1325145
pglazyfreed 0
pgrefill 0
pgsteal_kswapd 0
pgsteal_direct 0
pgscan_kswapd 0
pgscan_direct 0
pgscan_direct_throttle 0
pgscan_anon 0
pgscan_file 0
pgsteal_anon 0
pgsteal_file 0
zone_reclaim_failed 0
pginodesteal 0
slabs_scanned 0
kswapd_inodesteal 0
kswapd_low_wmark_hit_quickly 0
kswapd_high_wmark_hit_quickly 0
pageoutrun 0
pgrotated 0
drop_pagecache 0
drop_slab 0
oom_kill 0
numa_pte_updates 0
numa_huge_pte_updates 0
numa_hint_faults 0
numa_hint_faults_local 0
numa_pages_migrated 0
pgmigrate_success 0
pgmigrate_fail 0
compact_migrate_scanned 0
compact_free_scanned 0
compact_isolated 0
compact_stall 0
compact_fail 0
compact_success 0
compact_daemon_wake 0
compact_daemon_migrate_scanned 0
compact_daemon_free_scanned 0
htlb_buddy_alloc_success 0
htlb_buddy_alloc_fail 0
unevictable_pgs_culled 39356
unevictable_pgs_scanned 0
unevictable_pgs_rescued 64
unevictable_pgs_mlocked 64
unevictable_pgs_munlocked 64
unevictable_pgs_cleared 0
unevictable_pgs_stranded 0
thp_fault_alloc 23113
thp_fault_fallback 0
thp_fault_fallback_charge 0
thp_collapse_alloc 396
thp_collapse_alloc_failed 0
thp_file_alloc 0
thp_file_fallback 0
thp_file_fallback_charge 0
thp_file_mapped 0
thp_split_page 0
thp_split_page_failed 0
thp_deferred_split_page 1584
thp_split_pmd 20464
thp_split_pud 0
thp_zero_page_alloc 1
thp_zero_page_alloc_failed 0
thp_swpout 0
thp_swpout_fallback 0
balloon_inflate 0
balloon_deflate 0
balloon_migrate 0
swap_ra 0
swap_ra_hit 0
nr_unstable 0
```

And if I look at what kernel 4.18 can actually add to /proc/vmstat, `cache` is not one of those items: https://github.com/torvalds/linux/blob/v4.18/mm/vmstat.c#LL1108C37-L1108C37 - it's possible this is added by the ECS agent though; I did not check for that.

**Comment by bryan-aguilar** [comment]
I took at a look at this 
```
if stats.Memory.Stats != nil {
	m.MemoryUtilized = (*stats.Memory.Usage - stats.Memory.Stats["cache"]) / bytesInMiB
}
```
Turns out the issue here is instead the unchecked de-referencing of `stats.Memory.Usage`. I was able to replicate this in a unit test. Map access to a  non-present key should yield a zero value. I have noticed there area  lot of other unchecked de-references in that file. I should be able to work on some unit tests to test `nil` behavior a bit better. I will take this on and reference this issue when the PR is submitted upstream.
